TEST_APPS/testcases/nfc/test_self.py

```diff
@@ -0,0 +1,248 @@
+can be read back.
+'''
+@test_case(CreamSconeSelfTests)
+def test_nfc_write_long(self):
+    messageRep = 'thequickbrownfoxjumpedoverthelazydog' # repeating message written
+    textLength = 200       # 2K < x < 4K
+    # calculate actual message to compare to using the library
+    message = nfc_messages.make_textrecord( nfc_messages.repeat_string_to_length(messageRep, textLength))
+    expected_message = str(message)
+
+    self.nfc_command("dev1", "initnfc")
+    response = self.nfc_command("dev1", "iseeprom")
+    eeprom = response.parsed['iseeprom']
+    if eeprom:
+        self.logger.info("Target includes NFCEEPROM: %s" % eeprom)
+
+    self.nfc_command("dev1", "erase")
+    self.nfc_command("dev1", "writelong %d %s" % (textLength,messageRep))
+    response = self.nfc_command("dev1", "readmessage")
+    # assert that read the eeprom contents gives textlength bytes (including framing bytes which will vary)
+    asserts.assertEqual(len(response.parsed['nfcmessage']), len(expected_message))
+    i = 0
+    # assert that read the eeprom contents gives thequickbrownfoxjumpedoverthelazydog repeated in loop
+    while i < len(response.parsed['nfcmessage']):
+        asserts.assertEqual(response.parsed['nfcmessage'][i], ord(expected_message[i]))
+        i = i + 1
+
+'''
+check - Query supported protocols if we have a controller
+'''
+@test_case(CreamSconeSelfTests)
+def test_nfc_get_controller_protocols(self):
+    self.nfc_command("dev1", "initnfc")
+
+    response = self.nfc_command("dev1", "iseeprom")
+    eeprom = response.parsed['iseeprom']
+    if eeprom:
+        self.logger.info("Test ignore - target includes NFCEEPROM: %s" % eeprom)
+    else:
+        response = self.nfc_command("dev1", "getprotocols")
+        self.logger.info("Protocols = %s" % response.parsed['protocols'])
+
+
+'''
+check - Can set used protocols if we have a controller
+Note: Currently only support Typ4 tags in PN512 driver
+'''
+@test_case(CreamSconeSelfTests)
+def test_nfc_set_controller_protocols(self):
+    self.nfc_command("dev1", "initnfc")
+
+    response = self.nfc_command("dev1", "iseeprom")
+    eeprom = response.parsed['iseeprom']
+    if eeprom:
+        self.logger.info("Test ignore - target includes NFCEEPROM: %s" % eeprom)
+    else:
+        response = self.nfc_command("dev1", "setprotocols t1t")
+        response = self.nfc_command("dev1", "setprotocols t2t")
+        response = self.nfc_command("dev1", "setprotocols t3t")
+        response = self.nfc_command("dev1", "setprotocols isodep")
+        response = self.nfc_command("dev1", "setprotocols nfcdep")
+        response = self.nfc_command("dev1", "setprotocols t5t")
+        response = self.nfc_command("dev1", "setprotocols t1t t2t t3t isodep nfcdep t5t")
+
+'''
+check - SmartPoster URI forms are supported (in the test-app)
+'''
+@test_case(CreamSconeSelfTests)
+def test_nfc_check_smartposter_uri_forms(self):
+    def enum(**enums):
```

Comment:
why not import Enum? which python ver are we using?

Reply:
we should support 2.7, 3.5 - 3.7

Reply:
I've not run this under 3.X yet though, so I am going to install and verify there are no issues under 3.7 before I push to resolve the above issues the @paul-szczepanek-arm has pointed out in the Python.

Reply:
it's the other way round, enum is a python 3 thing but I noticed you import from enum elsewhere in this PR, does that work with 2.7?

Reply:
@ConradBraam If you like, you can rebase this PR to pick up some newly-enabled Py3 tests in Travis CI.